[v14 ready] Improve noise scaling implementation (#678)
SciML/Catalyst.jl: docs/src/catalyst_applications/advanced_simulations.md

## Scaling the noise magnitude in the chemical Langevin equations
When using the CLE to generate SDEs from a CRN, it can sometimes be desirable to

> scale the magnitude of the noise terms. Here, each reaction of the system generates a separate noise term in the CLE. If you require identical scaling for all reaction, the `@default_noise_scaling` option can be used. Else, you can supply a `noise_scaling` metadata for each individual reaction, describing how to scale the noise for that reaction.

Reviewer: ```suggestion
scale the magnitude of the noise terms. Here, each reaction of the system generates a separate noise term in the CLE. If you require identical scaling for all reactions, the `@default_noise_scaling` option can be used. Else, you can supply a `noise_scaling` metadata for each individual reaction, describing how to scale the noise for that reaction.
```

p_1 = [:k1 => 1.0, :k2 => 1.0]

sprob_1 = SDEProblem(rn_1, u0, tspan, p_1)
sol_1 = solve(sprob_1, ImplicitEM())
plot(sol_1; idxs = :X1, ylimit = (0.0, 20.0))
```
Here we can see that the $X$ concentration fluctuates around a steady state of $X≈10.0$.

> Next, we wish increase the amount of noise with by a factor 2. To do so, we use the `@default_noise_scaling` option, to which we provide the desired scaling 

Reviewer: ```suggestion
Next, we wish increase the amount of noise by a factor 2. To do so, we use the `@default_noise_scaling` option, to which we provide the desired scaling 
```

p_1 = [:k1 => 1.0, :k2 => 1.0]

sprob_1 = SDEProblem(rn_1, u0, tspan, p_1)

> sol_1 = solve(sprob_1, ImplicitEM())

Reviewer: Why switch from using the default solver?

plot(sol_1; idxs = :X1, ylimit = (0.0, 20.0))
```
Here we can see that the `X` concentration fluctuations around a steady state of *X≈10.0*.

Next, we wish to introduce a noise scaling parameter, `η`. This will scale the
noise magnitude so that for *η≈0.0* the system lacks noise (and its SDE
simulations are identical to its ODE simulations) and for *η≈1.0* noise is not
scaled (and SDE simulations are identical to as if no noise scaling was used).
Setting *η<1.0* will reduce noise and *η>1.0* will increase noise. The syntax
for setting a noise scaling parameter `η` is
```@example ex3
rn_2 = @reaction_network begin
    @noise_scaling_parameters η
    (k1,k2), X1 <--> X2
end
```

> The `@noise_scaling_parameter` option is equivalent to the `@parameters` option, but also designates subsequent parameter as a *noise scaling parameter*. *η* becomes a parameter of the system, and we can now modualte its value to scale simualtion noise:

Reviewer: ```suggestion
The `@noise_scaling_parameter` option creates one or more new parameters with additional metadata that allows Catalyst to know they represent a noise scaling. *η* becomes a parameter of the system, and we can now modulate its value to scale simulation noise:
```

```@example ex3
u0 = [:X1 => 10.0, :X2 => 10.0]
tspan = (0.0, 10.0)
p_2 = [:k1 => 1.0, :k2 => 1.0, :η => 0.1]

sprob_2 = SDEProblem(rn_2, u0, tspan, p_2)
sol_2 = solve(sprob_2, ImplicitEM())
plot(sol_2; idxs = :X1, ylimit = (0.0, 20.0))
```

> It is worth noting that in the CLE, nosie is tied to *reactions* (and not species, which is a common missperception). If only a single noise scaling parameter is given, it will scale the noise for all reaction. However, it is also possible to set several nosie scaling parameters, with each scaling the noise of a single reaction. Our model has two reactions (`X1 --> X2` and `X2 --> X1`) so we will use two noise scaling parameters (`η1` and `η2`):

Reviewer: ```suggestion
It is worth noting that in the CLE, nosie is tied to *reactions* (and not species, which is a common missperception). If only a single noise scaling parameter is given, it will scale the noise for all reactions. However, it is also possible to set several noise scaling parameters, with each scaling the noise of a single reaction. Our model has two reactions (`X1 --> X2` and `X2 --> X1`) so we will use two noise scaling parameters (`η1` and `η2`):
```

```@example ex3
rn_3 = @reaction_network begin
    @noise_scaling_parameters η1 η2
    (k1,k2), X1 <--> X2
end
u0 = [:X1 => 10.0, :X2 => 10.0]
tspan = (0.0, 10.0)
p_3 = [:k1 => 1.0, :k2 => 1.0, :η1 => 0.1, :η2 => 1.0]

sprob_3 = SDEProblem(rn_3, u0, tspan, p_3)
```

> Both the noise scaling parameters and the reaction are ordered (these orders can be seen by calling `reactions(rn_3)` and `noise_scaling_parameters(rn_3)`, respectively). The i'th noise scaling parameter scales the noise of the i'th reaction. Plotting the results, we see that we have less fluctuation than for the first simulation, but more as compared to the second one (which is as expected):

Reviewer: Relying on an ordering seems clunky and very much not what one should do, though I know it is how we did it in the past. It seems likely to cause problems too if one is composing / extending systems together as the ordering may change.
